File: MicroFrontend/appMicro/ios/App/App/plugins/DigitalWallet/utils/DigitalWalletErrorHandler.swift
//
//  DigitalWalletErrorHandler.swift
//  App
//
//  Created by Juan Carlos Pinzon on 12/04/24.
//

import Foundation
import Capacitor

public class DigitalWalletErrorHandler: Error {
    var message: String = ""
    var code: String = ""
    
    enum ErrorTypes: String {
        case invalidData
        case nullDigitalCard
        case unknownError
        case emptyWallet
        case tokensNotFound
        case deleteTokenError
        case resumeTokenError
        case suspendTokenError
        case canPushCardInApplePayError
        case digitalPlatformsError
        case encodeUtilError
        case searchDigitalCardError
    }
    
    private static let errorMap: [DigitalWalletErrorHandler.ErrorTypes: String] = [
            .invalidData: "The data is in an invalid format",
            .unknownError: "An unknown error occurred",
            .nullDigitalCard: "Digital card is null",
            .emptyWallet: "Empty Wallet",
            .tokensNotFound: "tokens not found",
            .deleteTokenError: "token not removed",
            .resumeTokenError: "token not activated",
            .suspendTokenError: "token not deactivated",
            .canPushCardInApplePayError: "error in isCardInApplePay",
            .digitalPlatformsError: "error in getDigitalPlatforms",
            .encodeUtilError: "error encode To string object whit encodeUtil",
            .searchDigitalCardError: "error find card Digital"
          ]
    
    init(_ error: ErrorTypes,_ errorDetail: String = "") {
        _init(error, errorDetail)
    }
    
    private func _init(_ error: ErrorTypes,_ errorDetail: String = "") {
        if let message = DigitalWalletErrorHandler.errorMap[error] {
          self.message = errorDetail != "" ? errorDetail : message
          self.code = error.rawValue
        }
    }
    
    static func reject(call: CAPPluginCall, error: ErrorTypes, errorDetail: String = "") {
        let err = DigitalWalletErrorHandler(error, errorDetail)
        err.rejectCall(call)
    }
    
    public func rejectCall(_ call: CAPPluginCall) {
        NSLog("HCESDK:AVV:Error(\(code)): \(message)")
        call.reject(message, code)
    }
}

File: MicroFrontend/appMicro/android/app/src/main/java/plugins/DigitalWallet/utils/DigitalWalletErrorHandler.java
package plugins.DigitalWallet.utils;

import com.getcapacitor.PluginCall;
import java.util.HashMap;
import java.util.Map;

public class DigitalWalletErrorHandler {

    private String message = "";
    private String code = "";

    public enum ErrorType {
        INVALID_DATA,
        NULL_DIGITAL_CARD,
        UNKNOWN_ERROR,
        EMPTY_WALLET,
        TOKENS_NOT_FOUND,
        DELETE_TOKEN_ERROR,
        RESUME_TOKEN_ERROR,
        SUSPEND_TOKEN_ERROR,
        CAN_PUSH_CARD_IN_GOOGLE_PAY_ERROR,
        DIGITAL_PLATFORMS_ERROR,
        ENCODE_UTIL_ERROR,
        SEARCH_DIGITAL_CARD_ERROR
    }

    private static final Map<ErrorType, String> errorMap = new HashMap<>() {{
        put(ErrorType.INVALID_DATA, "The data is in an invalid format");
        put(ErrorType.NULL_DIGITAL_CARD, "Digital card is null");
        put(ErrorType.UNKNOWN_ERROR, "An unknown error occurred");
        put(ErrorType.EMPTY_WALLET, "Empty Wallet");
        put(ErrorType.TOKENS_NOT_FOUND, "tokens not found");
        put(ErrorType.DELETE_TOKEN_ERROR, "token not removed");
        put(ErrorType.RESUME_TOKEN_ERROR, "token not activated");
        put(ErrorType.SUSPEND_TOKEN_ERROR, "token not deactivated");
        put(ErrorType.CAN_PUSH_CARD_IN_GOOGLE_PAY_ERROR, "error in isCardInGooglePay");
        put(ErrorType.DIGITAL_PLATFORMS_ERROR, "error in getDigitalPlatforms");
        put(ErrorType.ENCODE_UTIL_ERROR, "error encode To string object with encodeUtil");
        put(ErrorType.SEARCH_DIGITAL_CARD_ERROR, "error find card Digital");
    }};

    public DigitalWalletErrorHandler(ErrorType error, String errorDetail) {
        init(error, errorDetail);
    }

    public DigitalWalletErrorHandler(ErrorType error) {
        init(error, "");
    }

    private void init(ErrorType error, String errorDetail) {
        String defaultMessage = errorMap.getOrDefault(error, "Unknown error");
        this.message = (errorDetail != null && !errorDetail.isEmpty()) ? errorDetail : defaultMessage;
        this.code = error.name(); // enum name as string (like rawValue in Swift)
    }

    public static void reject(PluginCall call, ErrorType error, String errorDetail) {
        DigitalWalletErrorHandler err = new DigitalWalletErrorHandler(error, errorDetail);
        err.rejectCall(call);
    }

    public static void reject(PluginCall call, ErrorType error) {
        reject(call, error, "");
    }

    public void rejectCall(PluginCall call) {
        System.out.println("HCESDK:AVV:Error(" + code + "): " + message);
        call.reject(message, code);
    }
}
